File: crates/beet_ui/src/widgets/color_scheme.js
// Seeds the document color scheme before first paint to avoid a flash, then
// follows the OS preference until the user makes an explicit choice.
// `setColorScheme("light" | "dark")` persists an explicit choice and applies it.
// `LIGHT_SCHEME` / `DARK_SCHEME` are injected by the `ColorSchemeScript` widget.
(function () {
	const KEY = "color-scheme";
	const root = document.documentElement;
	const media = globalThis.matchMedia("(prefers-color-scheme: dark)");

	function apply(scheme) {
		root.classList.remove(LIGHT_SCHEME, DARK_SCHEME);
		root.classList.add(scheme === "dark" ? DARK_SCHEME : LIGHT_SCHEME);
	}

	function stored() {
		return globalThis.localStorage?.getItem(KEY);
	}

	// an explicit `?color-scheme=light|dark` query param wins and is persisted,
	// mirroring the CLI `--color-scheme` flag on the terminal target.
	function queryOverride() {
		const value = new URLSearchParams(globalThis.location?.search).get(KEY);
		return value === "light" || value === "dark" ? value : null;
	}

	globalThis.setColorScheme = function (scheme) {
		globalThis.localStorage?.setItem(KEY, scheme);
		apply(scheme);
	};

	// follow the OS until an explicit choice is stored
	media.addEventListener("change", () => {
		if (!stored()) apply(media.matches ? "dark" : "light");
	});

	const override = queryOverride();
	if (override) globalThis.setColorScheme(override);
	else apply(stored() ?? (media.matches ? "dark" : "light"));
})();

File: crates/beet_ui/src/widgets/color_scheme.rs
//! `ColorSchemeScript` widget — an inline `<script>` that seeds the document
//! `.light-scheme`/`.dark-scheme` class before first paint and persists the
//! user's explicit choice.
//!
//! Web target only — non-web targets carry the scheme class directly on an
//! ancestor element. Drop it into the document `<head>` (eg the `head` slot) so
//! the scheme is applied before the body paints, avoiding a flash of the wrong
//! theme.
use crate::style::material::classes;
use beet_core::prelude::*;

/// Emits the bundled color-scheme seed/persist script as an inline `<script>`.
///
/// The script reads the persisted choice (else the OS `prefers-color-scheme`),
/// toggling the matching class on `<html>`. `setColorScheme("light" | "dark")`
/// becomes available globally for a theme switcher.
#[template]
pub fn ColorSchemeScript() -> impl Bundle {
	// inject the shared class-name constants so the script and the style rules
	// stay in lockstep.
	let body = format!(
		"const LIGHT_SCHEME={:?},DARK_SCHEME={:?};\n{}",
		classes::LIGHT_SCHEME.as_selector(),
		classes::DARK_SCHEME.as_selector(),
		include_str!("./color_scheme.js"),
	);
	rsx! {
		<script>{body}</script>
	}
}
